File: clients/ios/FamilyAI/Core/Authentication/SessionManager.swift
import Foundation

public enum SessionRestoreResult: Equatable, Sendable {
  case needsPairing
  case authenticated(PersonalPortalContext)
  case offline, revoked
}
public actor SessionManager {
  private let gateway: GatewayClientProtocol
  private let credentials: CredentialStoreProtocol
  private let now: @Sendable () -> Date
  private var renewalTask: Task<EntrySessionCredential, Error>?
  public init(
    gateway: GatewayClientProtocol, credentials: CredentialStoreProtocol,
    now: @escaping @Sendable () -> Date = Date.init
  ) {
    self.gateway = gateway
    self.credentials = credentials
    self.now = now
  }
  public func restore() async throws -> SessionRestoreResult {
    guard let profile = try await credentials.gatewayProfile(),
      let device = try await credentials.deviceAuthorization()
    else { return .needsPairing }
    guard let storedSession = try await credentials.session() else {
      return try await restoreByRenewing(profile: profile, device: device)
    }
    let session: EntrySessionCredential
    if storedSession.expiresAt <= now() {
      do { session = try await renew(profile: profile, device: device) } catch {
        return try await mapRenewalFailure(error)
      }
    } else {
      session = storedSession
    }
    do {
      return .authenticated(
        try await gateway.fetchPortalContext(baseURL: profile.baseURL, session: session))
    } catch GatewayClientError.server(let code)
      where code == .entrySessionExpired || code == .entrySessionInvalid
    {
      do {
        let renewed = try await renew(profile: profile, device: device)
        return .authenticated(
          try await gateway.fetchPortalContext(baseURL: profile.baseURL, session: renewed))
      } catch { return try await mapRenewalFailure(error) }
    } catch GatewayClientError.server(.deviceRevoked) {
      try await credentials.clearDeviceAndSession()
      return .revoked
    } catch GatewayClientError.unreachable, GatewayClientError.timeout { return .offline }
  }
  public func validSession() async throws -> EntrySessionCredential {
    guard let profile = try await credentials.gatewayProfile(),
      let device = try await credentials.deviceAuthorization()
    else { throw GatewayClientError.server(.deviceAuthInvalid) }
    if let existing = try await credentials.session(), existing.expiresAt > now() {
      return existing
    }
    return try await renew(profile: profile, device: device)
  }
  public func logout() async throws {
    guard let profile = try await credentials.gatewayProfile(),
      let session = try await credentials.session()
    else {
      try await credentials.clearSession()
      return
    }

    do {
      _ = try await gateway.logout(
        baseURL: profile.baseURL,
        session: session
      )
      try await credentials.clearSession()
    } catch GatewayClientError.server(.entrySessionExpired),
      GatewayClientError.server(.entrySessionInvalid)
    {
      try await credentials.clearSession()
    } catch GatewayClientError.server(.deviceRevoked) {
      try await credentials.clearDeviceAndSession()
    } catch {
      throw error
    }
  }
  public func unbind() async throws {
    guard let profile = try await credentials.gatewayProfile(),
      let device = try await credentials.deviceAuthorization()
    else {
      try await credentials.clearDeviceAndSession()
      return
    }

    do {
      _ = try await gateway.unbind(
        baseURL: profile.baseURL,
        authorization: device
      )
    } catch GatewayClientError.server(.deviceRevoked) {
      // The server already reached the desired terminal state.
    } catch {
      throw error
    }

    try await credentials.clearDeviceAndSession()
  }
  private func restoreByRenewing(profile: GatewayProfile, device: DeviceAuthorization) async throws
    -> SessionRestoreResult
  {
    do {
      let session = try await renew(profile: profile, device: device)
      return .authenticated(
        try await gateway.fetchPortalContext(baseURL: profile.baseURL, session: session))
    } catch { return try await mapRenewalFailure(error) }
  }
  private func renew(profile: GatewayProfile, device: DeviceAuthorization) async throws
    -> EntrySessionCredential
  {
    if let renewalTask { return try await renewalTask.value }
    let task = Task { [gateway, credentials] in
      let response = try await gateway.renew(baseURL: profile.baseURL, authorization: device)
      try await credentials.replaceSessionAtomically(response.entry)
      return response.entry
    }
    renewalTask = task
    defer { renewalTask = nil }
    return try await task.value
  }
  private func mapRenewalFailure(_ error: Error) async throws -> SessionRestoreResult {
    if case GatewayClientError.server(.deviceRevoked) = error {
      try await credentials.clearDeviceAndSession()
      return .revoked
    }
    if case GatewayClientError.unreachable = error { return .offline }
    if case GatewayClientError.timeout = error { return .offline }
    throw error
  }
}

File: clients/harmonyos/core/src/sessionManager.ts
import type { MobileCredentialStore } from './credentialStore.ts';
import {
  GatewayClientError,
  type GatewayProfile
} from './gatewayClient.ts';
import type { DeviceAuthorization } from './requests.ts';
import type {
  EntrySessionCredential,
  MobileOperationResponse,
  PersonalPortalContext,
  SessionRenewResponse
} from './types.ts';

export interface SessionGateway {
  fetchPortalContext(
    profile: GatewayProfile,
    session: EntrySessionCredential
  ): Promise<PersonalPortalContext>;
  renew(
    profile: GatewayProfile,
    authorization: DeviceAuthorization
  ): Promise<SessionRenewResponse>;
  logout(
    profile: GatewayProfile,
    authorization: DeviceAuthorization
  ): Promise<MobileOperationResponse>;
  unbind(
    profile: GatewayProfile,
    authorization: DeviceAuthorization
  ): Promise<MobileOperationResponse>;
}

export type SessionRestoreResult =
  | { kind: 'needsPairing' }
  | { kind: 'authenticated'; context: PersonalPortalContext }
  | { kind: 'offline' }
  | { kind: 'revoked' };

export class SessionManager {
  private renewalTask: Promise<EntrySessionCredential> | null = null;

  constructor(
    private readonly gateway: SessionGateway,
    private readonly credentials: MobileCredentialStore,
    private readonly now: () => number = () => Date.now()
  ) {}

  async restore(): Promise<SessionRestoreResult> {
    const profile = await this.credentials.gatewayProfile();
    const authorization = await this.credentials.deviceAuthorization();
    if (!profile || !authorization) return { kind: 'needsPairing' };

    const storedSession = await this.credentials.session();
    let activeSession: EntrySessionCredential;
    if (!storedSession || Date.parse(storedSession.expiresAt) <= this.now()) {
      try {
        activeSession = await this.renew(profile, authorization);
      } catch (error) {
        return this.mapRestoreFailure(error);
      }
    } else {
      activeSession = storedSession;
    }

    try {
      const context = await this.gateway.fetchPortalContext(profile, activeSession);
      return { kind: 'authenticated', context };
    } catch (error) {
      if (this.isEntrySessionInvalid(error)) {
        try {
          const renewed = await this.renew(profile, authorization);
          const context = await this.gateway.fetchPortalContext(profile, renewed);
          return { kind: 'authenticated', context };
        } catch (renewalError) {
          return this.mapRestoreFailure(renewalError);
        }
      }
      return this.mapRestoreFailure(error);
    }
  }

  async validSession(): Promise<EntrySessionCredential> {
    const profile = await this.credentials.gatewayProfile();
    const authorization = await this.credentials.deviceAuthorization();
    if (!profile || !authorization) {
      throw new GatewayClientError('server', 'DEVICE_AUTH_INVALID');
    }

    const existing = await this.credentials.session();
    if (existing && Date.parse(existing.expiresAt) > this.now()) return existing;
    return this.renew(profile, authorization);
  }

  async logout(): Promise<void> {
    const profile = await this.credentials.gatewayProfile();
    const authorization = await this.credentials.deviceAuthorization();
    if (!profile || !authorization) {
      await this.credentials.clearSession();
      return;
    }

    try {
      await this.gateway.logout(profile, authorization);
      await this.credentials.clearSession();
    } catch (error) {
      if (this.isServerCode(error, 'DEVICE_REVOKED')) {
        await this.credentials.clearDeviceAndSession();
        return;
      }
      throw error;
    }
  }

  async unbind(): Promise<void> {
    const profile = await this.credentials.gatewayProfile();
    const authorization = await this.credentials.deviceAuthorization();
    if (!profile || !authorization) {
      await this.credentials.clearDeviceAndSession();
      return;
    }

    try {
      await this.gateway.unbind(profile, authorization);
    } catch (error) {
      if (!this.isServerCode(error, 'DEVICE_REVOKED')) throw error;
    }
    await this.credentials.clearDeviceAndSession();
  }

  private async renew(
    profile: GatewayProfile,
    authorization: DeviceAuthorization
  ): Promise<EntrySessionCredential> {
    if (this.renewalTask) return this.renewalTask;

    const task = (async () => {
      const response = await this.gateway.renew(profile, authorization);
      await this.credentials.replaceSessionAtomically(response.entry);
      return response.entry;
    })();
    this.renewalTask = task;

    try {
      return await task;
    } finally {
      if (this.renewalTask === task) this.renewalTask = null;
    }
  }

  private async mapRestoreFailure(error: unknown): Promise<SessionRestoreResult> {
    if (this.isServerCode(error, 'DEVICE_REVOKED')) {
      await this.credentials.clearDeviceAndSession();
      return { kind: 'revoked' };
    }
    if (
      error instanceof GatewayClientError &&
      (error.kind === 'unreachable' || error.kind === 'timeout')
    ) {
      return { kind: 'offline' };
    }
    throw error;
  }

  private isEntrySessionInvalid(error: unknown): boolean {
    return this.isServerCode(error, 'ENTRY_SESSION_EXPIRED') ||
      this.isServerCode(error, 'ENTRY_SESSION_INVALID');
  }

  private isServerCode(error: unknown, code: GatewayClientError['serverCode']): boolean {
    return error instanceof GatewayClientError &&
      error.kind === 'server' &&
      error.serverCode === code;
  }
}
